### PR TITLE
refactor: Compiler config in Rust

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes https://github.com/denoland/deno/issues/5606

### CODE REVIEW

File: cli/tsc.rs [hunk 1]
```diff
@@ -288,18 +292,24 @@ impl CompilerConfig {
...
+    } else {
+      None
+    };
+
     // If `checkJs` is set to true in `compilerOptions` then we're gonna be compiling
     // JavaScript files as well
-    let compile_js = if let Some(config_content) = config.clone() {
-      let config_str = std::str::from_utf8(&config_content)?;
+    let compile_js = if let Some(config_str) = content.as_ref() {
       CHECK_JS_RE.is_match(config_str)
```
Comment (kitsonk) [inline cli/tsc.rs:305]: we should really be finding a way to refactor this out... Some like this based on the JSON value:

```rust
let mut compile_js = false;
if let Some(value) = compiler_options.as_object().unwrap().get("checkJs") {
  compile_js = value.as_bool().unwrap_or(false);
}
```

Comment (bartlomieju) [reply cli/tsc.rs:305]: No worries, will be done, yesterday I was mainly fighting to keep CI green 👍

Comment (bartlomieju) [reply cli/tsc.rs]: Done

**Comment by bartlomieju** [comment]
@kitsonk all configuration has been moved to Rust (`cli/tsc.rs`) - I kept the logic exactly as was in `cli/tsc/99_main_compiler.js` - I guess it can be simplified with next passes. 

Please review and let me know if this patch is satisfactory.

CC @ry also please review

File: cli/msg.rs [hunk 2]
```diff
@@ -55,7 +55,6 @@ pub fn enum_name_media_type(mt: MediaType) -> &'static str {
 #[derive(Clone, Copy, PartialEq, Debug)]
 pub enum CompilerRequestType {
   Compile = 0,
-  Transpile = 1,
```
Comment (kitsonk) [inline cli/msg.rs:58]: Seems unrelated to this PR (and now we have a hole in the sequence)

File: cli/tsc.rs [hunk 3]
```diff
@@ -203,6 +202,16 @@ lazy_static! {
     Regex::new(r#"(\slib\s*=\s*)('|")(.+?)('|")"#).unwrap();
 }
 
+fn warn_ignored_options(
+  maybe_ignored_options: Option<tsc_config::IgnoredCompilerOptions>,
+  config_path: &Path,
+) {
+  if let Some(ignored_options) = maybe_ignored_options {
+    let ignored_opts = ignored_options.to_string();
```
Comment (kitsonk) [inline cli/tsc.rs:207]: Doesn't this happen automatically on formatting?  It should.

File: cli/tsc.rs [hunk 4]
```diff
@@ -726,33 +731,62 @@ impl TsCompiler {
...
+      maybe_ignored_options,
+      compiler_config.path.as_ref().unwrap(),
+    );
+
+    let j = json!({
+      "type": msg::CompilerRequestType::Bundle,
+      "target": target,
+      "rootNames": root_names,
+      "performance": performance,
+      "configPath": compiler_config.path.unwrap_or_else(|| PathBuf::from("tsconfig.json")),
```
Comment (kitsonk) [inline cli/tsc.rs]: We can in theory just drop this and fake it always... it simply means that we don't support project references or extends, which we never supported anyways, and if we do, we would do it in Rust anyways and provide a single set of options anyways.

This is what I did in the rewrite branch: https://github.com/kitsonk/deno/blob/deno_compiler/compiler/compiler.js#L187-L194

File: cli/tsc/99_main_compiler.js [hunk 5]
```diff
@@ -1484,77 +1427,11 @@ delete Object.prototype.__proto__;
     };
   }
 
-  function transpile({
```
Comment (kitsonk) [inline cli/tsc/99_main_compiler.js:1455]: Again, I would do it as a seperate PR...

Comment (bartlomieju) [reply cli/tsc/99_main_compiler.js:1455]: Opened https://github.com/denoland/deno/pull/7277, I'll rebase this branch after landing.

File: cli/tsc.rs (hunk 4 above)
Comment (bartlomieju) [reply cli/tsc.rs]: I tried, see comment in `parseCompilerOptions` method. I agree we should remove it completely.

File: cli/tsc.rs (hunk 3 above)
Comment (bartlomieju) [reply cli/tsc.rs:207]: It does, thanks

File: cli/tsc.rs (hunk 4 above)
Comment (bartlomieju) [reply cli/tsc.rs]: Actually already did

File: cli/tsc_config.rs [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+  assert!(!config_text.as_ref().is_empty());
+  let jsonc = jsonc_parser::parse_to_value(config_text.as_ref())?.unwrap();
+  Ok(jsonc_to_serde(jsonc))
+}
+
+/// Take a string of JSONC, parse it and return a serde `Value` of the text.
+/// The result also contains any options that were ignored.
+pub fn parse_config<T: AsRef<str>>(
+  config_text: T,
+) -> Result<(Value, Option<IgnoredCompilerOptions>), ErrBox> {
```
Comment (ry) [inline cli/tsc_config.rs:154]: This is a stylistic nitpick but I prefer `&str` arguments over `AsRef<str>`... I think in some places it makes sense to be more generic but here there is just a single caller.

File: cli/tsc_config.rs [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+  extends: Option<String>,
+  files: Option<Vec<String>>,
+  include: Option<Vec<String>>,
+  references: Option<Value>,
+  type_acquisition: Option<Value>,
+}
+
+pub fn parse_raw_config<T: AsRef<str>>(
+  config_text: T,
+) -> Result<Value, ErrBox> {
```
Comment (ry) [inline cli/tsc_config.rs:144]: ditto https://github.com/denoland/deno/pull/7228/files#r480208812

File: cli/tsc_config.rs (hunk 6 above)
Comment (bartlomieju) [reply cli/tsc_config.rs:154]: 👍  sure, done
